compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantValueResolver.java

```diff
@@ -234,8 +237,10 @@ public void visit(BLangRecordLiteral recordLiteral) {
                 } else {
                     continue;
                 }
-
+                BType tempCurrentConstSymbolType = this.currentConstSymbolType;
+                this.currentConstSymbolType = keyValuePair.getBType();
                 value = constructBLangConstantValue(keyValuePair.valueExpr);
+                this.currentConstSymbolType = tempCurrentConstSymbolType;
```

Comment:
- We should deal with the global variables very carefully and avoid as much as possible in the compiler as one wrong assignment could mess up the whole compiler.

- Keeping two separate global variables for `constant.symbol` and `constant.symbol.type` anyway is not a good practice, as updating one place wouldn't update the other place in parallel. (i.e. updating `currentConstSymbolType` would not update `currentConstSymbol`'s type)

- Here, temporarily setting the value `this.currentConstSymbolType = keyValuePair.getBType();` is hacky and would be specific to this scenario only wouldn't solve this for non-record scenarios. 

As I understand the root cause of this issue is as follows.

At the `ConstantValueResolver` we expect to invoke the `void visit(BLangConstant constant) {}` visitor first, where we initialize the global variables. However, in this scenario, we invoke the `ConstantValueResolver` via the `constructBLangConstantValueWithExactType()` which bypass the said visit. 

The correct fix is, we need to initialize the `currentConstSymbol` at the beginning of `constructBLangConstantValueWithExactType()`. i.e. Need to add `this.currentConstSymbol = constantSymbol;` line.